### DISCUSSION TITLE
Enforce the convention that None is not allowed as a node?

### DISCUSSION THREAD

**Discussion by dschult** [networkx/networkx#4788] [Ideas]
This discussion stems from a comment in #4757 wondering if it would be worth enforcing the convention that None is not allowed as a node.  How much would that affect the codebase (difficult to implement?) and the performance (does it create lots of time checking).

I think it would simply require a single check for each newly added node.  This occurs in methods `add_node` and `add_nodes_from` and probably in `add_edge` and `add_edges_from` and possibly one or two other base class methods where performance dictated manually adding the node -- maybe we got rid of all those already.  

I don't think it would require any other additional checks -- just the "main front door" -- but that should be verified.

**Comment by rossbar** [comment]
FWIW I tried adding `raise ValueError` to `Graph.add_node` just to see what the impact would be on the test suite and that did cause a test to fail, so there is at least one test that violates the `None can't be a node` rule.

This is the type of change that it would be really nice to have a benchmarking suite in place for, since it's a change to the core data structure. I don't *think* adding checks to raise an exception would really slow things down that much, but it would be nice to have hard evidence :)

**Comment by dschult** [comment]
We had a benchmark.py file up until 2017 which I ran occasionally to see if we were regressing. I dug it up from git and ran it to test `add_nodes_from` and `add_edges_from`. The times with and without a check for None raising a ValueError are at the 3rd significant digit. Well within the spread between runs of the tests. And that's a harsh test where all we time is adding the node.  

So, using `if node is None:` apparently doesn't slow things down at all. 

The broken test is in `test_function.py` and slipped through to get merged in #2759 even though it uses `None` as a node.  We should correct that erroneous test (maybe use True as the test node instead.) at the same time that we add tests that `add_node` and `add_edge` raise an error...

**Comment by rossbar** [comment]
> So, using if node is None: apparently doesn't slow things down at all.

Great - that's what I would've expected, but it's nice to have some more quantitative evidence. Thanks for checking it out!

**Comment by boothby** [comment]
Why _is_ `None` forbidden as a node label?  Will allowing `None` as a node label introduce nasty bugs in algorithms that depend on this invariant?  A quick grep for None shows almost 3k instances in the codebase.  But, I do find uses such as

```
networkx/drawing/nx_pydot.py:def pydot_layout(G, prog="neato", root=None):
networkx/drawing/nx_pydot.py-    """Create node positions using :mod:`pydot` and Graphviz.
...
networkx/drawing/nx_pydot.py:    root : Node from G or None (default: None)
networkx/drawing/nx_pydot.py-        The node of G from which to start some layout algorithms.
...
networkx/drawing/nx_pydot.py:    if root is not None:
networkx/drawing/nx_pydot.py-        P.set("root", str(root))
```

An obvious solution would be to add an object `null_node = object()` to the top-level namespace, and use that in place of `None` for cases like this.  But I can't think of a way to cleanly automate that process, and 3k hits in grep is going to be quite a slog.

Oh, sorry... this was pre-coffee.  I see now that we're discussing the enforcement, and not the removal.  Count this as a vote for enforcement.

**Comment by dschult** [comment]
I created a PR to enforce None is not allowed as a node.
